### DISCUSSION TITLE
Is there a way for JDT to delegate the compilation request in the IDE to a plugin?

### DISCUSSION THREAD

**Discussion by laeubi** [eclipse-jdt/eclipse.jdt.core#2372] [Q&A]
Currently with m2e we have the situation that inside the IDE we must disable the maven-compiler-plugin because JDT is performing the compile.

This has several pitfalls, because in Maven one can configure many things more flexible than in JDT (e.g. different executions with different settings for different source folders (see for example https://github.com/eclipse-jdt/eclipse.jdt.core/issues/265) but the main issue is that compilation works different in the IDE than if one executes it on the maven cli. 

I therefore aim to something like that whenever JDT is detecting that a source file needs to be recompiled, it does not perform this directly but ask a extension / service / ... to do the actual work and gets some callback to inform JDT when the new class file is available.

**Comment by iloveeclipse** [comment]
Switch off Java builder on the project and add a custom one that does whatever you want.

**Reply by laeubi** [reply]
How does JDT then knows that something has changed? I got the impression that java builder is doing more than just the compile e.g. triggers building related projects on change, update indexes, ...

**Reply by iloveeclipse** [reply]
> How does JDT then knows that something has changed?

The new builder should properly use IResources API I would say :-)

> I got the impression that java builder is doing more than just the compile e.g. triggers building related projects on change, update indexes, ...

You have to discover what works and what not. I can't say without looking in the code which parts are triggered by the default build infra and which are by the Java builder.
